### DISCUSSION TITLE
Feature Request: Notification when changes directly applied to K8s

### DISCUSSION THREAD

**Discussion by steveww** [fluxcd/flux2#2334] [Proposals]
With GitOps all changes should be via Git. If a user (accidentally or maliciously) edits a K8s resource, Flux will revert the change back to the single source of truth. At the moment the event notification is the same as when Flux applies a change as a result of a commit to Git. The events should be different or have additional context.

There are many cases when it is desirable to act on reconciliation events such as:

* Notify DevSecOps
* Cert manager installing new cert

**Comment by stefanprodan** [comment]
> Cert manager installing new cert

Flux applies the certificate custom resource, then cert-manager generates a secret in cluster. The generated secret is not in Git so Flux doesn't override it. You may want to raise this issue in cert-manager repo, it's not in Flux scope to notify users of another controller's actions.

**Comment by steveww** [comment]
Point taken on Cert Manager.

The other use case is still not covered.

**Reply by kingdonb** [reply]
Certs can be stored in git, as long as there is a workflow for recognizing when they are about to be signed again for renewal and committing the resulting certificate before it's overwritten again by the next syncing interval. I do use Flux this way, and I'm not here to argue the point, because another tool like Vault is honestly better suited to this job right now, it might not always be the case.

It's not very convenient TBH as it is now, but it works and saving certs and keys in Git provides a way around the Let's Encrypt rate limit for clusters that are likely to be destroyed and re-created more than 5 times per week. As long as you have a way of monitoring certs for expiry and renewal dates upcoming, which I'm 50% on now, because [only nginx-ingress seems to provide good monitoring](https://github.com/kingdonb/flux2/blob/32e56b0d8a455db587f3f7b3e867f2bee4b7198c/manifests/monitoring/grafana/dashboards/nginx.json#L1312) of that, (and only possible as far as I'm aware via an extension to Grafana dashboards)...

Then users can suspend reconciling of the `certificates` Kustomization and force a renewal, or come back to collect the renewed certificates from the cluster before re-enabling the suspended Kustomization. This is kludgy and manual; if I was able to detect and alert on changes in a deployed resource that wasn't reflected in the git source, then I could take action before Flux does and remediate the issue with a commit.

This "works" now but it would be much better with a real mechanism for alerting on unexpected drift.

Now that (https://github.com/fluxcd/flux2/pull/2167) has landed, with `flux diff` I think it will be possible for users to build out these types of workflows on their own, hopefully released soon (starting with Flux 0.26.) 🎉 That was a main missing feature to make it work.

It will be tricky to make these alerts general enough so they can handle races and distinguish between changes that are in queue to apply, changes that are currently applying, and changes that have already been applied, all could be detected and shown as distinct cases from "drifts that Flux expected to encounter" aka changes that came through Git channels. I can think of a number of simple ways to implement this that would not account for those races, and so I would not recommend it (as this is likely to be hooked up to a pager, alerts should be rock solid and unlikely to generate false positives, or they will cause users to distrust the alerting systems.) There are of course likely better ways than I imagined!
